common/service/dynamicconfig/config.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamicconfig
+
+import (
+	"errors"
+	"time"
+)
+
+// Client allows fetching values from a dynamic configuration system NOTE: This does not have async
+// options right now. In the interest of keeping it minimal, we can add when requirement arises.
+type Client interface {
+	GetValue(name Key) (interface{}, error)
+	GetValueWithFilters(name Key, filters map[Filter]interface{}) (interface{}, error)
+}
+
+type nopClient struct{}
+
+func (mc *nopClient) GetValue(name Key) (interface{}, error) {
+	return nil, errors.New("unable to find key")
+}
+
+func (mc *nopClient) GetValueWithFilters(
+	name Key, filters map[Filter]interface{},
+) (interface{}, error) {
+	return nil, errors.New("unable to find key")
+}
+
+// NewNopCollection creates a new nop collection
+func NewNopCollection() *Collection {
+	return NewCollection(&nopClient{})
+}
+
+// NewCollection creates a new collection
+func NewCollection(client Client) *Collection {
+	return &Collection{client}
+}
+
+// Collection of values that are in dynamic config
+type Collection struct {
+	client Client
+}
+
+// GetIntPropertyWithTaskList gets property with taskList filter and asserts that it's an integer
+func (c *Collection) GetIntPropertyWithTaskList(key Key, defaultVal int) func(string) int {
```

Comment:
I would make this package non Cadence specific. So this method would be something like:

```
func (c *Collection) GetIntPropertyWithStringFilter(key Key, defaultValue int, filter Filter) func(string) int {
	return func(filterValue string) int {
 		filters := make(map[Filter]interface{})
 		filters[filter] = filterValue
 		val, err := c.client.GetValueWithFilters(key, filters)
 		if err != nil {
 			return defaultVal
 		}
 		return val.(int)
 	}
 }
```
Then the code that uses it would make it task list specific. If the task list appears as filter in a lot of places the 
then the following function (outside of this package) can be created:

```
func GetIntPropertyWithTaskListFilter(c *Collection, key Key, defaultValue int) func(string) int {
      return c.GetIntPropertyWithString(key, defaultValue, TaskListName)
}
```

Reply:
What's the benefit of a public generic function? Makes sense as a private method if we have public methods for both task list and domain name.
Also all these methods are for Cadence, what is a non-Cadence use case for this?